### PR TITLE
feat: payment processor token for recurring payments

### PR DESCRIPTION
**Pull request by prajjwalkumar17**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

During the transition phase of Hyperswitch, Merchant should be able to send the processor's token to Hyperswitch to process MIT transaction. This is because, Merchant will not have the card data to migrate from Processor to Hyperswitch.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

### The testing flows are as follows:
1. Create a Setup Mandate Payment.
2. Get the payment_method_id.
3. Get the connector_mandate_id from payment_method table in db (using pm_method_id) -> this is our payment_processor_token (PPT).
4. Now create a recurring_payment using the above ppt.

The curls for the same are as follows:
1. Create a Setup Mandate Payment.
```
curl --location 'http://127.0.0.1:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: ••••••' \
--data-raw '
{
    "amount": 10000,
    "currency": "USD",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "customer_id": "cus_35RDuRN4P6XdVMgt2Px7",
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "setup_future_usage": "off_session",
    "off_session": true,
    "browser_info": {
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "127.0.0.1"
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "John",
            "last_name": "Doe"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "payment_type": "setup_mandate",
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "02",
            "card_exp_year": "26",
            "card_holder_name": "Joseph Doe",
            "card_cvc": "999"
        }
    },
    "billing": {
        "address": {
            "line1": "8th block",
            "line2": "8th block",
            "line3": "8th block",
            "city": "Bengaluru",
            "state": "Karnataka",
            "zip": "560095",
            "country": "IN",
            "first_name": "Sakil",
            "last_name": "Mostak"
        }
    },
    "customer_acceptance": {
        "acceptance_type": "online"
    }
}'
```
2. Get the payment_method_id.
```
"payment_method_id": "pm_A1YI0BNBELMQVw76ipT3",
```

3. Get the connector_mandate_id from payment_method table in db (using pm_method_id) -> this is our payment_processor_token (PPT).
```
connector_mandate_details      | {"mca_A3fAZBZEG6X8cBb19z9w": {"payment_method_type": "credit", "connector_mandate_id": "**pm_1Pkk1YD5R7gDAGff2yPGAPgr**", "original_payment_authorized_amount": 10000, "original_payment_authorized_currency": "USD"}}
```

4. Now create a recurring_payment using the above ppt (confirm true).
```
curl --location 'http://127.0.0.1:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_QP5rWb7dq1xSRpxdQoroFUxUiD3kLx6MgfsBBEe9FlvkAjXnqTYINyylTR8o5zkG' \
--data '{
    "amount": 499,
    "currency": "USD",
    "capture_method": "automatic",
    "customer_id": "cus_35RDuRN4P6XdVMgt2Px7",
    "off_session": true,
    "confirm": true,
    "recurring_details": {
        "type": "processor_payment_token",
        "data": {
            "processor_payment_token": "pm_1Pkk1YD5R7gDAGff2yPGAPgr",
            "connector": "stripe",
            "merchant_connector_id": "mca_A3fAZBZEG6X8cBb19z9w"
        }
    },
    "payment_method": "card",
    "authentication_type": "no_three_ds"
}'
```

4. Now create a recurring_payment using the above ppt (confirm false).
```
curl --location 'http://127.0.0.1:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_QP5rWb7dq1xSRpxdQoroFUxUiD3kLx6MgfsBBEe9FlvkAjXnqTYINyylTR8o5zkG' \
--data '{
    "amount": 499,
    "currency": "USD",
    "capture_method": "automatic",
    "customer_id": "cus_35RDuRN4P6XdVMgt2Px7",
    "off_session": true,
    "confirm": false,
    "authentication_type": "no_three_ds"
}'
```

5. Confirm the payment
```
curl --location 'http://localhost:8080/payments/pay_LG5wWKTQ5sLATRBchdc3/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: ••••••' \
--data '{
    "off_session": true,
    "payment_method": "card",
    "recurring_details": {
        "type": "processor_payment_token",
        "data": {
            "processor_payment_token": "pm_1Pkk1YD5R7gDAGff2yPGAPgr",
            "connector": "stripe",
            "merchant_connector_id": "mca_A3fAZBZEG6X8cBb19z9w"
        }
    }
}'
```

## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/payments/operations/payment_approve.rs [hunk 1]
```diff
@@ -181,6 +181,7 @@ impl<F: Send + Clone> GetTracker<F, PaymentData<F>, api::PaymentsCaptureRequest>
             authentication: None,
             recurring_details: None,
             poll_config: None,
+            is_payment_processor_token_flow: None,
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_approve.rs:184]: we do not need this field in payment data as we already have payment intent which has this field, use that instead

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 2]
```diff
@@ -628,14 +628,39 @@ impl<F: Send + Clone> GetTracker<F, PaymentData<F>, api::PaymentsRequest> for Pa
             .as_ref()
             .map(|payment_method_billing| payment_method_billing.address_id.clone());
 
+        // If processor_payment_token is passed in request then populating the same in PaymentData
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_confirm.rs:631]: we do not need this in payments confirm

File: crates/router/src/core/payments/operations/payment_create.rs [hunk 3]
```diff
@@ -376,6 +376,32 @@ impl<F: Send + Clone> GetTracker<F, PaymentData<F>, api::PaymentsRequest> for Pa
             .await
             .transpose()?;
 
+        let mandate_id = if mandate_id.is_none() {
+            request
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_create.rs:380]: can you move this match to the place where it is previously being matched? Move this to a function maybe?

File: crates/router/src/core/payments/operations/payment_confirm.rs (hunk 2 above)
Comment (prajjwalkumar17) [reply crates/router/src/core/payments/operations/payment_confirm.rs:631]: As discussed, For future usecases we will let this flow be present in confirm as well.

File: crates/router/src/core/payments/operations/payment_create.rs (hunk 3 above)
Comment (prajjwalkumar17) [reply crates/router/src/core/payments/operations/payment_create.rs:380]: Marking this as resolved as per our discussion.

**Review by Narayanbhat166 — APPROVED** [review]

File: crates/api_models/src/mandates.rs [hunk 4]
```diff
@@ -120,4 +120,14 @@ pub struct MandateListConstraints {
...
+    ProcessorPaymentToken(ProcessorPaymentToken),
+}
+
+/// Processor payment token for MIT payments where payment_method_data is not available
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, ToSchema)]
+pub struct ProcessorPaymentToken {
+    pub processor_payment_token: String,
+    #[schema(value_type = Connector, example = "stripe")]
+    pub connector: api_enums::Connector,
+    pub merchant_connector_id: String,
```
Comment (SanchithHegde) [inline crates/api_models/src/mandates.rs:132]: Can we document the fields as well?

**Review by SanchithHegde — APPROVED** [review]

**Pull request merged**